tensorlayer/db.py

```diff
@@ -5,11 +5,725 @@
+            _result_key = task['result_key']
+            logging.info("  hyper parameters:")
+            for key in _hyper_parameters:
+                globals()[key] = _hyper_parameters[key]
+                logging.info("    {}: {}".format(key, _hyper_parameters[key]))
+            ## run task
+            # f = open('_ztemp.py', 'wb')
+            # f.write(task['script'])
+            # f.close()
+            s = time.time()
+            logging.info("[Database] Start Task: key: {} sort: {} push time: {}".format(task_key, sort, _datetime))
+            # os.system("python __ztemp.py")
+            # import _ztemp
+            _script = _script.decode('utf-8')
+            with tf.Graph().as_default():  # as graph: # clear all TF graphs
+                exec(_script, globals())
+            # os.remove("_ztemp.py")
+
+            ## set status to finished
+            _ = self.db.Task.find_one_and_update({'_id': _id}, {'$set': {'status': 'finished'}})
+
+            ## return results
+            __result = {}
+            for _key in _result_key:
+                logging.info("  result: {}={} {}".format(_key, globals()[_key], type(globals()[_key])))
+                # print(type(_key), type(globals()[_key]))
+                # __result__.update({_key, globals()[str(_key)]})
+                __result.update({"%s" % _key: globals()[_key]})
+            # print(__result, type(globals()[_key]))
+            _ = self.db.Task.find_one_and_update(
+                {
+                    '_id': _id
+                }, {'$set': {
+                    'result': __result
+                }}, return_document=pymongo.ReturnDocument.AFTER
+            )
+            # print(_['result'])
+            logging.info(
+                "[Database] Finished Task: task_key - {} sort: {} push time: {} took: {}s".format(
+                    task_key, sort, _datetime,
+                    time.time() - s
+                )
+            )
+            # exit()
+            return True
+        except Exception as e:
+            exc_type, exc_obj, exc_tb = sys.exc_info()
+            fname = os.path.split(exc_tb.tb_frame.f_code.co_filename)[1]
+            logging.info("{}  {}  {}  {}  {}".format(exc_type, exc_obj, fname, exc_tb.tb_lineno, e))
+            logging.info("[Database] Fail to run task")
+            ## if fail, set status back to pending
+            _ = self.db.Task.find_one_and_update({'_id': _id}, {'$set': {'status': 'pending'}})
+            return False
+
+    def del_task(self, **kwargs):
+        """Delete tasks.
+
+        Parameters
+        -----------
+        kwargs : logging information
+            Find items to delete, leave it empty to delete all log.
+
+        Examples
+        ---------
+        >>> db.del_task()
+        """
+        self._fill_project_info(kwargs)
+        self.db.Task.delete_many(kwargs)
+        logging.info("[Database] Delete Task SUCCESS")
+
+    def check_unfinished_task(self, task_key=None, **kwargs):
```

Comment:
if you already gives the task_key, what is the deference between "find_task"

Reply:
this one is for task distributor check whether all tasks finished.